source/adapters/level_zero/program.cpp

```diff
@@ -601,12 +601,16 @@ UR_APIEXPORT ur_result_t UR_APICALL urProgramGetInfo(
     return ReturnValue(uint32_t{Program->RefCount.load()});
   case UR_PROGRAM_INFO_CONTEXT:
     return ReturnValue(Program->Context);
-  case UR_PROGRAM_INFO_NUM_DEVICES:
-    // TODO: return true number of devices this program exists for.
-    return ReturnValue(uint32_t{1});
-  case UR_PROGRAM_INFO_DEVICES:
-    // TODO: return all devices this program exists for.
-    return ReturnValue(Program->Context->Devices[0]);
+  case UR_PROGRAM_INFO_NUM_DEVICES: {
+    return ReturnValue(static_cast<uint32_t>(Program->ZeModuleMap.size()));
+  }
+  case UR_PROGRAM_INFO_DEVICES: {
+    std::vector<ze_device_handle_t> devices;
+    for (const auto &entry : Program->ZeModuleMap) {
+      devices.push_back(entry.first);
+    }
+    return ReturnValue(devices.data(), devices.size());
+  }
```

Comment:
I think that should report the devices inside the context. From what I could find `ZeModuleMap` will only be accumulated when either `urProgramBuildExp` is called or `urProgramLinkExp` and we could call this query before calling these entries (I might be missing something here).

Reply:
Yea, this was wrong :D My fault for rushing.
From what I can see, all paths to build a program go through the exp functions except for urProgramCreateWithBinary. In which case `ZeModuleMap` is going to be empty.

Thanks.